File: smarthome-back/mqtt_client/ev_charger.go
package mqtt_client

import (
	"encoding/json"
	"fmt"
	mqtt "github.com/eclipse/paho.mqtt.golang"
	influxdb2 "github.com/influxdata/influxdb-client-go/v2"
	"smarthome-back/dtos"
	"strconv"
	"strings"
	"time"
)

func (mc *MQTTClient) HandleInputPercentageChange(client mqtt.Client, msg mqtt.Message) {
	// Retrieve the last part of the split string, which is the device ID
	parts := strings.Split(msg.Topic(), "/")
	deviceId, err := strconv.Atoi(parts[len(parts)-1])
	if err != nil {
		fmt.Println(err)
	}

	// Unmarshal the JSON string into the struct
	var data dtos.ElectricVehicleDTO
	err = json.Unmarshal([]byte(msg.Payload()), &data)
	if err != nil {
		fmt.Println("Error unmarshaling JSON:", err)
		return
	}
	fmt.Println(data.Email)
	saveActionToInflux(mc.influxDb, deviceId, data.Email, "percentageChange", -1, data.CurrentCapacity)
}

func (mc *MQTTClient) HandleAutoActionsForCharger(client mqtt.Client, msg mqtt.Message) {
	// Retrieve the last part of the split string, which is the device ID
	parts := strings.Split(msg.Topic(), "/")
	deviceId, err := strconv.Atoi(parts[len(parts)-1])
	fmt.Println(deviceId)
	if err != nil {
		fmt.Println(err)
	}

	var data dtos.ElectricVehicleDTO
	err = json.Unmarshal([]byte(msg.Payload()), &data)
	if err != nil {
		fmt.Println("Error unmarshaling JSON:", err)
		return
	}

	saveActionToInflux(mc.influxDb, deviceId, "auto", data.Action, data.PlugId, data.CurrentCapacity)
}

// actions can be start, end and percentageChange
func saveActionToInflux(client influxdb2.Client, deviceId int, user string, action string, plugId int, percentage float64) {
	Org := "Smart Home"
	Bucket := "bucket"
	writeAPI := client.WriteAPI(Org, Bucket)
	p := influxdb2.NewPoint("ev_charger", //table
		map[string]string{"device_id": strconv.Itoa(deviceId), "user_id": user, "plug_id": strconv.Itoa(plugId), "action": action}, //tag
		map[string]interface{}{"value": percentage}, //field
		time.Now())

	// Write the point to InfluxDB
	writeAPI.WritePoint(p)

	// Close the write API to flush the buffer and release resources
	writeAPI.Flush()
	fmt.Printf("Savin to influxdb. Electrical charger: id=%d, action %s, plugId %d, percentage %f \n", deviceId, action, plugId, percentage)

}

File: scripts/student3/ev_charger.py
import math
import random
from datetime import datetime, timedelta
from influxdb_client import InfluxDBClient, Point, WritePrecision
from influxdb_client.client.write_api import ASYNCHRONOUS

INFLUX_TOKEN = "5JrdbOYRvz9k03QXYOEO5LvHkB9pBVDRimwkr26IDRd9aDDvsHxBno4H3e412w7DZ--Rk8Ltgny2MURH4OUd5A=="
ORG = "Smart Home"
BUCKET = "bucket"

url = "http://localhost:8086"
influxdb_client = InfluxDBClient(url=url, token=INFLUX_TOKEN, org=ORG)


def generate_data_ev_charger(deviceId, start_time, end_time):
    write_api = influxdb_client.write_api(write_options=ASYNCHRONOUS)
    current_time = start_time
    actions = ['start', 'end', 'percentageChange']

    while current_time <= end_time:
        action = random.choice(actions)
        plug_id = random.randint(0, 1)
        percentage = random.uniform(0.2, 0.9)
        if action != 'percentageChange':
            user = 'auto'
        else:
            user = 'nata@gmail.com'
        point = (
            Point("ev_charger")
            .tag("device_id", deviceId)
            .tag("user_id", user)
            .tag("plug_id", plug_id)
            .tag("action", action)
            .field("value", percentage)
            .time(current_time, WritePrecision.NS)
        )

        write_api.write(bucket=BUCKET, org=ORG, record=point)
        print(f"Written point: {point} at {current_time}")

        # Increment time by 12 hours
        current_time += timedelta(hours=12)


if __name__ == "__main__":
    # Define the start and end time for the data
    end_time = datetime.utcnow()
    start_time = end_time - timedelta(days=90)  # 3 months ago

    generate_data_ev_charger(13, start_time, end_time)

    # Close the client
    influxdb_client.close()
